### PR TITLE
[WIP] bf16_mkldnn_gemm

### PR DESCRIPTION
**Pull request by zhuhaozhe**
### Changes for Mkldnn-linear (with mkldnn inner product pritimitve):
https://oneapi-src.github.io/oneDNN/group__dnnl__api__inner__product.html
- Input tensor type:
    - Before: input must be mkldnn tensor
    - Now: input can be a bf16 cpu tensor (Do not convert bf16 cpu tensor to bf16 mkldnn tensor can save a mem copy)
- Weight packs:
    - Before: to_mkldnn(model) only convert nn.Linear to nn.mkldnn_linear (weight will not be packed)
    - Now: weight will be packed after converted to mkldnn-linear to avoid run-time format conversion on each iter

### User Case1:
```
import torch
LL = torch.nn.Linear(256, 256).bfloat16()
# forward
LL(torch.rand(16, 256).bfloat16())
```
### User Case2 (preferred):
```
import torch
LL = torch.nn.Linear(256, 256).bfloat16()
mkldnn_LL = mkldnn_utils.to_mkldnn(LL, dtype=torch.bfloat16) # will pack weight from plain format to expected block format to accelerate gemm 
# forward
mkldnn_LL(torch.rand(16, 256).bfloat16())
```

### BF16 training:
For training, we suppose the batch size is big, so the reorder of weight for each iter will not be a big overhead.
```
import torch
LL = torch.nn.Linear(256, 256).bfloat16()
# forward
y = LL(torch.rand(16, 256).bfloat16())
# backward
y.mean().backward()
```

## Performance Gain for Linear
Note: data is collected on 28core/per socket CPX with using pytorch operator benchmarks: https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark (with adding bfloat dtype and mkldnn-linear benchmark for prepacking weight). The backward bench also contains autograd engine part.
### Forward
| impl | dtype |shape | #threads/cores | time(us)
| :---: | :---: | :---: | :---: | :---: |
| aten | fp32 |16, 512, 256 | 1 |  63.314 |
| aten | bf16 |16, 512, 256 | 1 | 10127.247|
| mkldnn | bf16 |16 , 512 , 256 | 1 |  102.591| 
| mkldnn_packed_weight | bf16 |16 , 512 , 256 | 1 | 41.916|
| aten | fp32 |16 , 512 , 256 | 28 | 26.595 |
| aten | bf16 |16 , 512 , 256 | 28 |  10051.861| 
| mkldnn | bf16 |16 , 512 , 256 | 28 | 54.086 |
| mkldnn_packed_weight | bf16 |16 , 512 , 256 | 28 | 43.295|
| aten | fp32 |1,1,1 | 1 | 9.119|
| aten | bf16 |1,1,1 | 1 | 9.128|
| mkldnn | bf16 |1,1,1 | 1 | 29.920| 
| mkldnn_packed_weight | bf16 |1,1,1 | 1 |  27.915|
| aten | fp32 |1,1,1 | 28 | 9.899 |
| aten | bf16 |1,1,1 | 28 | 9.258| 
| mkldnn | bf16 |1,1,1 | 28 | 31.030 |
| mkldnn_packed_weight | bf16 |1,1,1 | 28 | 28.952|
### Backward
| impl | dtype |shape | #threads/cores | time(us)
| :---: | :---: | :---: | :---: | :---: |
| aten | fp32 |2048, 512, 256 | 28|  424.863 |
| aten | bf16 |2048, 512, 256 | 28 | 240046.316|
| mkldnn | bf16 |2048 , 512 , 256 | 28 | 1854.447| 

When shapes are quiet small, gemm computation dose not takes large proportion of time costs. So we get performance gain when  shapes are big enough.

### BF16 Suport for mm, addmm, bmm, addbmm, baddbmm (with mkldnn matmul primitive):
https://oneapi-src.github.io/oneDNN/group__dnnl__api__matmul.html
For gemm related ops, we keep all inputs under plain format. So we will not introduce opaque tensor for these ops to save mem copy here.
mkldnn bf16 gemm integration

The minimized integration is only dispatch to mkldnn in addmm, but for gemm with 3-D input (with additional dim for"batch") this will call mkldnn gemm for "batch" times. Since mkldnn matmul support input with multiple dims, we directly dispatch to mkldnn gemm in {bmm, addbmm, baddbmm} to reduce the time to create mkldnn memory desc, primitive, etc.

For the different definition for "bias" between mkldnn(which must be shape of (1, N)) and pytorch (which can be same shape with gemm result (M, N)), we use a fused sum to handle it.

### User Case:
User case is exactly same with before since no opaque tensor's is introduced. Since the pytorch has already support bf16 data type with CPU tensor before, we can leverage the existed bf16 gemm UT.
### torch.mm
| impl | dtype |shape | #threads/cores | time(us)|
|:---:| :---: | :---: | :---: | :---: |
| aten | fp32 |1, 1, 1 | 1|  2.134 |
| aten | bf16 |1, 1, 1 | 1 | 1.925|
| mkldnn | bf16 |1 , 1 , 1 | 1 | 19.039| 
| aten | fp32 |128, 256, 512 | 1|  158.602 |
| aten | bf16 |128, 256, 512 | 1 | 15176.579|
| mkldnn | bf16 |128, 256, 512 | 1 |  140.450 | 
| aten | fp32 |1, 1, 1 | 28|  2.324 |
| aten | bf16 |1, 1, 1 | 28 | 2.074|
| mkldnn | bf16 |1 , 1 , 1 | 28 | 24.775| 
| aten | fp32 |128, 256, 512 | 28|  40.707 |
| aten | bf16 |128, 256, 512 | 28 | 15203.331|
| mkldnn | bf16 |128, 256, 512 | 28 | 55.462 | 

### torch.addmm
| impl | dtype |shape | #threads/cores | time(us)|
|:---:| :---: | :---: | :---: | :---: |
| aten | fp32 |1, 1, 1 | 1|  2.898 |
| aten | bf16 |1, 1, 1 | 1 | 2.648|
| mkldnn | bf16 |1 , 1 , 1 | 1 |  20.344| 
| aten | fp32 |128, 256, 512 | 1|  184.459 |
| aten | bf16 |128, 256, 512 | 1 | 15184.096|
| mkldnn | bf16 |128, 256, 512 | 1 |   146.364 | 
| aten | fp32 |1, 1, 1 | 28|   2.955 |
| aten | bf16 |1, 1, 1 | 28 | 2.707|
| mkldnn | bf16 |1 , 1 , 1 | 28 |  26.236| 
| aten | fp32 |128, 256, 512 | 28|  59.082 |
| aten | bf16 |128, 256, 512 | 28 | 15952.905|
| mkldnn | bf16 |128, 256, 512 | 28 |  56.245| 

### torch.addbmm
| impl | dtype |shape | #threads/cores | time(us)|
|:---:| :---: | :---: | :---: | :---: |
| aten | fp32 |1, 1, 1, 1 | 1|  4.152 |
| aten | bf16 |1, 1, 1, 1 | 1 | 3.827|
| mkldnn | bf16 |1, 1 , 1 , 1 | 1 | 21.434 | 
| aten | fp32 |8, 128, 256, 512 | 1|  1335.469 |
| aten | bf16 |8, 128, 256, 512 | 1 | 121503.256|
| mkldnn | bf16 |8, 128, 256, 512 | 1 |883.417 | 
| aten | fp32 |1, 1, 1, 1 | 28|   4.263|
| aten | bf16 |1, 1, 1, 1 | 28 |   3.781 |
| mkldnn | bf16 |1, 1 , 1 , 1 | 28 |  27.775  | 
| aten | fp32 |8, 128, 256, 512 | 28|  384.288|
| aten | bf16 |8, 128, 256, 512 | 28 | 1122196.027|
| mkldnn | bf16 |8, 128, 256, 512 | 28 | 144.470| 

### torch.bmm
| impl | dtype |shape | #threads/cores | time(us)|
|:---:| :---: | :---: | :---: | :---: |
| aten | fp32 |1, 1, 1, 1 | 1|  1.811 |
| aten | bf16 |1, 1, 1, 1 | 1 | 1.973|
| mkldnn | bf16 |1, 1 , 1 , 1 | 1 | - | 
| aten | fp32 |8, 128, 256, 512 | 1|  1270.769 |
| aten | bf16 |8, 128, 256, 512 | 1 | 121749.807|
| mkldnn | bf16 |8, 128, 256, 512 | 1 | 896.570 | 
| aten | fp32 |1, 1, 1, 1 | 28|   1.854 |
| aten | bf16 |1, 1, 1, 1 | 28 |  1.905 |
| mkldnn | bf16 |1, 1 , 1 , 1 | 28 |  - | 
| aten | fp32 |8, 128, 256, 512 | 28|  97.950 |
| aten | bf16 |8, 128, 256, 512 | 28 | 121695.516|
| mkldnn | bf16 |8, 128, 256, 512 | 28 | 169.577| 

### torch.baddbmm
| impl | dtype |shape | #threads/cores | time(us)|
|:---:| :---: | :---: | :---: | :---: |
| aten | fp32 |1, 1, 1, 1 | 1|  2.613 |
| aten | bf16 |1, 1, 1, 1 | 1 | 2.738|
| mkldnn | bf16 |1, 1 , 1 , 1 | 1 | - | 
| aten | fp32 |8, 128, 256, 512 | 1|  1438.951 |
| aten | bf16 |8, 128, 256, 512 | 1 | 121985.955|
| mkldnn | bf16 |8, 128, 256, 512 | 1 | 942.380 | 
| aten | fp32 |1, 1, 1, 1 | 28|    2.556 |
| aten | bf16 |1, 1, 1, 1 | 28 |  2.703 |
| mkldnn | bf16 |1, 1 , 1 , 1 | 28 |  - | 
| aten | fp32 |8, 128, 256, 512 | 28|  120.843 |
| aten | bf16 |8, 128, 256, 512 | 28 | 122489.345|
| mkldnn | bf16 |8, 128, 256, 512 | 28 | 168.992| 

For bmm and baddbmm (B, M, N) * (B, N, K),  if (M * N * K) <= 400 will follow original dispatch

### CODE REVIEW

**Comment by zhuhaozhe** [comment]
@jgong5 @mingfeima @XiaobingSuper Please help to review.

**Comment by ezyang** [comment]
closed?

**Comment by jgong5** [comment]
@ezyang It was split into https://github.com/pytorch/pytorch/pull/61339 and https://github.com/pytorch/pytorch/pull/64619. The latter was landed.
